docs/design/design.md

```diff
@@ -0,0 +1,261 @@
+## Implementation design for kubeadm
+
+`kubeadm init` and `kubeadm join` together provides a nice user experience for creating a best-practice but bare Kubernetes cluster from scratch.
+However, it might not be obvious _how_ kubeadm does that.
+
+This document strives to explain what types of phases of work happen under the hood.
+Also included is ComponentConfiguration API types for talking to kubeadm programmatically.
```

Comment:
No, it's what I think the new ComponentConfiguration stuff looks like: https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/apis/kubeadm/v1alpha1/types.go

in it's own API group and everything. Not coupled to anything else, just reusing the API machinery